stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4135,65 +4155,20 @@ LogicalResult TransposeOp::reifyReturnTypeShapes(
   return success();
 }
 
-// Method for InferTypeOpInterface: infer the return type from the operand type
-// and the permutation.
 LogicalResult TransposeOp::inferReturnTypes(
     MLIRContext* context, Optional<Location> loc, ValueRange operands,
-    DictionaryAttr attributes, RegionRange,
+    DictionaryAttr attributes, RegionRange regions,
     SmallVectorImpl<Type>& inferredReturnTypes) {
-  auto type = operands[0].getType();
-  auto rankedTy = type.dyn_cast<RankedTensorType>();
-  if (!rankedTy) {
-    auto shapedTy = type.dyn_cast<ShapedType>();
-    inferredReturnTypes.emplace_back(shapedTy);
-    return success();
-  }
-  auto permutation = attributes.getAs<DenseIntElementsAttr>("permutation");
-  int64_t rank = rankedTy.getRank();
-  if (permutation.getType().getRank() != 1)
-    return emitOptionalError(loc, "TransposeOp permutation has rank ",
-                             permutation.getType().getRank(),
-                             " instead of rank 1");
-
-  if (permutation.size() != rank)
-    return emitOptionalError(loc, "TransposeOp operand rank ", rank,
-                             " does not match permutation size ",
-                             permutation.size());
-
-  std::vector<int64_t> range(rank);
-  std::iota(range.begin(), range.end(), 0);
-  if (!std::is_permutation(range.begin(), range.end(), permutation.begin()))
-    return emitOptionalError(loc,
-                             "attribute permutation must be a permutation"
-                             " of [",
-                             range, "] but got ", permutation);
-
-  llvm::SmallVector<int64_t, 4> inputBounds(rank, ShapedType::kDynamicSize);
-  bool hasBounds = false;
-  if (auto encoding =
-          rankedTy.getEncoding().dyn_cast_or_null<TypeExtensionsAttr>()) {
-    inputBounds = llvm::to_vector<4>(encoding.getBounds());
-    hasBounds = true;
-  }
+  TransposeOp::Adaptor adaptor(operands, attributes, regions);
+  auto inputBounds = getBounds(adaptor.getOperand().getType());
+  llvm::SmallVector<int64_t> resultBounds;
 
-  SmallVector<int64_t> resultShape;
-  SmallVector<int64_t> resultBounds;
-  ArrayRef<int64_t> inputShape = rankedTy.getShape();
-  for (int64_t dim : permutation.getValues<int64_t>()) {
-    resultShape.push_back(inputShape[dim]);
-    if (hasBounds) {
-      resultBounds.push_back(inputBounds[dim]);
-    }
-  }
+  LogicalResult result = hlo::inferTransposeOp(
+      context, loc, adaptor.getOperand(), inputBounds, adaptor.getPermutation(),
+      inferredReturnTypes, resultBounds);
 
-  // If the input type doesn't have bounds, propagate the input type's encoding
-  // to handle sparse tensor encoding.
```

Comment:
"Does the existing code https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/StablehloOps.cpp#L4189 mean that we can't have bounds and sparse tensor encoding at the same time?". Yes, that's right.

Reply:
What if we need both? I guess it is not possible until this TODO is done? https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/StablehloAttrs.td#L134

Reply:
I agree. I don't think this will be possible until this TODO is resolved or we find some other way to combine the two.